File: 15-3sum/15-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        HashSet<List<Integer>> list = new HashSet<>();
        HashMap<Integer, Integer> map = new HashMap();
        for(int i=0; i<nums.length; i++) {
            map.put(nums[i], i);
        }
        for(int i=0; i<nums.length; i++) {
            int z = nums[i];
            if(z > 0) break;
            for(int j=i+1; j<nums.length; j++) {
                if(i == j) continue;
                int x = nums[j];
                int y = x + z;
                if(y > 0) break;
                y = -1 * y;
                if(map.containsKey(y) && map.get(y) > i && map.get(y) > j) {
                    list.add(List.of(x, y, z));
                }
            }
        }
        return new ArrayList(list);
    }
}

File: 15-3sum/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        
        res = []
        
        i=0
        
        while(i < (len(nums))):
            if(nums[i] > 0): break
            l = i+1
            r = len(nums)-1
            while(l < r):
                sum = nums[i] + nums[l] + nums[r]
                
                if(sum > 0):
                    r -= 1
                elif(sum < 0):
                    l += 1
                else:
                    lis = [nums[i], nums[l], nums[r]]
                    res.append(lis)
                    while(l+1 < len(nums) and nums[l] == nums[l+1]):
                        l += 1
                    l += 1
                    r -= 1
            while(i+1 < len(nums) and nums[i] == nums[i+1]):
                i += 1
            i += 1
        
        return res